web/ui/react-app/src/pages/graph/ExpressionInput.tsx

```diff
@@ -169,7 +175,15 @@ const ExpressionInput: FC<CMExpressionInputProps> = ({
             ])
           ),
           EditorView.updateListener.of((update: ViewUpdate): void => {
-            onExpressionChange(update.state.doc.toString());
+            setHasLinterErrors(diagnosticCount(view.state) > 0);
+
+            if (update.docChanged) {
+              onExpressionChange(update.state.doc.toString());
+
+              if (exprFormatted) {
+                setExprFormatted(false);
```

Comment:
also you could simply do

```typescript
onExpressionChange(update.state.doc.toString());
setExprFormatted(false);
```

Reply:
Yes, the only (weak) reason for the `if` condition is to not cause yet another React state update when it's not needed (if `exprFormatted` is already `false`, we don't need to set it to `false`), since this happens on every typed character. But yeah, probably it doesn't matter at all. I also tried to do the same with the `setHasLinterErrors()`, but it led to weird feedback issues due to the lag, so now I'm just always setting it.